File: LeetCode/199. Binary Tree Right Side View/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

# class Solution:
#     def rightSideView(self, root: TreeNode) -> List[int]:
#         q=collections.deque()
#         if root==None: return []
#         q.appendleft(root)
#         ans=[]
#         cur_count=1
#         next_count=0
#         while cur_count>0:
#             node=q.pop()
            
#             if node.left!=None:
#                 q.appendleft(node.left)
#                 next_count+=1
#             if node.right!=None:
#                 q.appendleft(node.right)
#                 next_count+=1
#             if cur_count==1:
#                 ans.append(node.val)
#                 cur_count=next_count
#                 next_count=0
#             else:
#                 cur_count-=1

#         return ans

from collections import deque
class Solution:
    def rightSideView(self, root: TreeNode) -> List[int]:
        if root is None:
            return []
        results=[]
        q=deque()
        q.append(root)
        while len(q)>0:
            # bfs
            results.append(q[-1].val)
            for _ in range(len(q)):
                n=q.popleft()
                if n.left is not None:
                    q.append(n.left)
                if n.right is not None:
                    q.append(n.right)
        return results
                

class Solution:
    def rightSideView(self, root: TreeNode) -> List[int]:
        if not root:
            return []
        q = [root]
        res = []
        while q:
            print(q)
            res.append(q[-1].val)
            tmp = []
            for cur in q:
                if cur.left:
                    tmp.append(cur.left)
                if cur.right:
                    tmp.append(cur.right)
            q = tmp
        return res

File: LeetCode/199. Binary Tree Right Side View/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
        vector<int> res;
        vector<TreeNode*> q, nq;
        if (root) q.push_back(root);
        while (!q.empty()){
            int n = q.size();
            nq.clear();
            for (int i = 0; i < n; i++){
                TreeNode *cur = q[i];
                if (i == n-1) res.push_back(cur->val);
                if (cur->left) nq.push_back(cur->left);
                if (cur->right) nq.push_back(cur->right);
            }
            swap(q,nq);
        }
        return res;
    }
};